File: beautifulplanet/safepaw/services/router/internal/consumer/consumer.go
// =============================================================
// NOPEnclaw Router — Redis Streams Consumer
// =============================================================
// Reads messages from nopenclaw_inbound using XREADGROUP.
//
// Why XREADGROUP instead of XREAD:
// 1. Consumer groups: multiple Router instances share the load.
//    Each message goes to exactly ONE consumer (not all of them).
// 2. Acknowledgment: XACK confirms processing. If the Router
//    crashes mid-process, unacked messages can be reclaimed.
// 3. Pending list: Redis tracks which messages are in-flight.
//    We can detect stuck messages and retry them.
//
// This is the difference between "demo that breaks at scale"
// and "system that handles failures gracefully."
// =============================================================

package consumer

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	goredis "github.com/redis/go-redis/v9"

	"nopenclaw/router/internal/config"
)

// Message represents a message read from the inbound stream.
// Includes the stream ID needed for acknowledgment.
type Message struct {
	StreamID       string            `json:"-"`
	MessageID      string            `json:"message_id"`
	SessionID      string            `json:"session_id"`
	Channel        string            `json:"channel"`
	SenderID       string            `json:"sender_id"`
	SenderPlatform string            `json:"sender_platform"`
	ContentType    string            `json:"content_type"`
	Content        string            `json:"content"`
	Metadata       map[string]string `json:"metadata,omitempty"`
	Timestamp      int64             `json:"timestamp"`
	// Retry tracking
	DeliveryCount int64 `json:"-"`
}

// Handler is a function that processes a single message.
// Returns an error if processing fails (message will be retried).
type Handler func(ctx context.Context, msg *Message) error

// Consumer reads from a Redis Stream using consumer groups.
type Consumer struct {
	client *goredis.Client
	cfg    *config.Config
	wg     sync.WaitGroup

	// work is a buffered channel that distributes messages to workers.
	// This decouples the read loop from processing, preventing a slow
	// message from blocking reads of new messages.
	work chan *Message
}

// NewConsumer creates a Consumer and ensures the consumer group exists.
// If the group already exists, this is a no-op (idempotent).
func NewConsumer(cfg *config.Config) (*Consumer, error) {
	client := goredis.NewClient(&goredis.Options{
		Addr:         cfg.RedisAddr,
		Password:     cfg.RedisPassword,
		DB:           cfg.RedisDB,
		PoolSize:     cfg.WorkerCount + 5, // Workers + reader + health + headroom
		MinIdleConns: 2,
		MaxRetries:   3,
		DialTimeout:  5 * time.Second,
		ReadTimeout:  time.Duration(cfg.BlockTime.Milliseconds()+2000) * time.Millisecond, // Block time + buffer
		WriteTimeout: 3 * time.Second,
		PoolTimeout:  4 * time.Second,
	})

	// Verify connection
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		client.Close()
		return nil, fmt.Errorf("redis connection failed: %w", err)
	}

	// Create consumer group if it doesn't exist.
	// "0" means: start reading from the beginning of the stream.
	// MKSTREAM creates the stream if it doesn't exist yet.
	err := client.XGroupCreateMkStream(ctx, cfg.InboundStream, cfg.ConsumerGroup, "0").Err()
	if err != nil {
		// "BUSYGROUP" means group already exists — that's fine
		if err.Error() != "BUSYGROUP Consumer Group name already exists" {
			client.Close()
			return nil, fmt.Errorf("failed to create consumer group %q: %w", cfg.ConsumerGroup, err)
		}
	}

	log.Printf("[CONSUMER] Connected to %s, group=%s, consumer=%s, workers=%d",
		cfg.RedisAddr, cfg.ConsumerGroup, cfg.ConsumerName, cfg.WorkerCount)

	return &Consumer{
		client: client,
		cfg:    cfg,
		work:   make(chan *Message, cfg.WorkerCount*2), // 2× workers for buffering
	}, nil
}

// Run starts the consumer loop and worker pool. Blocks until ctx is cancelled.
func (c *Consumer) Run(ctx context.Context, handler Handler) {
	// Start worker pool
	for i := 0; i < c.cfg.WorkerCount; i++ {
		c.wg.Add(1)
		go c.worker(ctx, i, handler)
	}

	// Start pending message reclaimer (recovers crashed messages)
	c.wg.Add(1)
	go c.pendingReclaimer(ctx)

	// Main read loop — reads NEW messages from the stream
	c.readLoop(ctx)

	// Shutdown: close work channel, wait for workers to finish
	close(c.work)
	c.wg.Wait()
}

// readLoop continuously reads new messages from the stream.
func (c *Consumer) readLoop(ctx context.Context) {
	log.Printf("[CONSUMER] Read loop started (batch=%d, block=%v)",
		c.cfg.BatchSize, c.cfg.BlockTime)

	for {
		select {
		case <-ctx.Done():
			log.Println("[CONSUMER] Read loop stopping (context cancelled)")
			return
		default:
		}

		// XREADGROUP GROUP <group> <consumer> COUNT <n> BLOCK <ms> STREAMS <stream> >
		// The ">" means: only give me messages that haven't been delivered to any consumer yet
		results, err := c.client.XReadGroup(ctx, &goredis.XReadGroupArgs{
			Group:    c.cfg.ConsumerGroup,
			Consumer: c.cfg.ConsumerName,
			Streams:  []string{c.cfg.InboundStream, ">"},
			Count:    c.cfg.BatchSize,
			Block:    c.cfg.BlockTime,
		}).Result()

		if err != nil {
			if ctx.Err() != nil {
				return // Shutting down
			}
			if err == goredis.Nil {
				continue // Timeout, no messages — normal
			}
			log.Printf("[CONSUMER] XREADGROUP error: %v (retrying in 1s)", err)
			time.Sleep(1 * time.Second)
			continue
		}

		for _, stream := range results {
			for _, xmsg := range stream.Messages {
				msg, err := parseMessage(xmsg)
				if err != nil {
					log.Printf("[CONSUMER] Malformed message %s: %v (acking to skip)", xmsg.ID, err)
					// ACK malformed messages so they don't clog the pending list
					c.ack(ctx, xmsg.ID)
					continue
				}

				// Send to worker pool. If workers are all busy and buffer is full,
				// this blocks — which naturally applies backpressure to Redis reads.
				select {
				case c.work <- msg:
				case <-ctx.Done():
					return
				}
			}
		}
	}
}

// worker processes messages from the work channel.
func (c *Consumer) worker(ctx context.Context, id int, handler Handler) {
	defer c.wg.Done()
	log.Printf("[WORKER-%d] Started", id)

	for msg := range c.work {
		if ctx.Err() != nil {
			return
		}

		// Process with timeout
		processCtx, cancel := context.WithTimeout(ctx, c.cfg.AckTimeout)
		err := handler(processCtx, msg)
		cancel()

		if err != nil {
			log.Printf("[WORKER-%d] Processing failed for msg=%s: %v (delivery #%d)",
				id, msg.MessageID, err, msg.DeliveryCount)
			// Don't ACK — message stays in pending list for retry by pendingReclaimer
			continue
		}

		// Success — acknowledge the message
		c.ack(ctx, msg.StreamID)
	}

	log.Printf("[WORKER-%d] Stopped", id)
}

// pendingReclaimer periodically checks for stuck messages (delivered but not acked)
// and re-queues them for processing.
//
// This handles the case where:
// - A worker crashes mid-processing
// - A message takes too long and the worker moves on
// - A previous Router instance died with unacked messages
func (c *Consumer) pendingReclaimer(ctx context.Context) {
	defer c.wg.Done()

	ticker := time.NewTicker(c.cfg.AckTimeout) // Check at the same interval as the timeout
	defer ticker.Stop()

	log.Printf("[RECLAIMER] Started (check interval=%v, max retries=%d)",
		c.cfg.AckTimeout, c.cfg.MaxRetries)

	for {
		select {
		case <-ctx.Done():
			log.Println("[RECLAIMER] Stopped")
			return
		case <-ticker.C:
			c.reclaimPending(ctx)
		}
	}
}

// reclaimPending checks the pending list for stuck messages.
func (c *Consumer) reclaimPending(ctx context.Context) {
	// XPENDING <stream> <group> - + <count>
	// Returns messages that were delivered but not yet acknowledged
	pending, err := c.client.XPendingExt(ctx, &goredis.XPendingExtArgs{
		Stream: c.cfg.InboundStream,
		Group:  c.cfg.ConsumerGroup,
		Start:  "-",
		End:    "+",
		Count:  c.cfg.BatchSize,
		Idle:   c.cfg.AckTimeout, // Only messages idle longer than the timeout
	}).Result()

	if err != nil {
		if ctx.Err() != nil {
			return
		}
		log.Printf("[RECLAIMER] XPENDING error: %v", err)
		return
	}

	if len(pending) == 0 {
		return
	}

	log.Printf("[RECLAIMER] Found %d stuck messages", len(pending))

	for _, p := range pending {
		if ctx.Err() != nil {
			return
		}

		// Check retry count — if exceeded, dead-letter the message
		if p.RetryCount >= int64(c.cfg.MaxRetries) {
			log.Printf("[RECLAIMER] Message %s exceeded max retries (%d/%d) — dead-lettering",
				p.ID, p.RetryCount, c.cfg.MaxRetries)
			// ACK to remove from pending, log for manual investigation
			// In production, this would go to a dead-letter stream
			c.ack(ctx, p.ID)
			continue
		}

		// XCLAIM: take ownership of the stuck message and re-deliver it
		claimed, err := c.client.XClaim(ctx, &goredis.XClaimArgs{
			Stream:   c.cfg.InboundStream,
			Group:    c.cfg.ConsumerGroup,
			Consumer: c.cfg.ConsumerName,
			MinIdle:  c.cfg.AckTimeout,
			Messages: []string{p.ID},
		}).Result()

		if err != nil {
			log.Printf("[RECLAIMER] XCLAIM failed for %s: %v", p.ID, err)
			continue
		}

		for _, xmsg := range claimed {
			msg, err := parseMessage(xmsg)
			if err != nil {
				log.Printf("[RECLAIMER] Malformed reclaimed message %s: %v (acking)", xmsg.ID, err)
				c.ack(ctx, xmsg.ID)
				continue
			}
			msg.DeliveryCount = p.RetryCount + 1

			select {
			case c.work <- msg:
				log.Printf("[RECLAIMER] Re-queued message %s (attempt #%d)", msg.MessageID, msg.DeliveryCount)
			case <-ctx.Done():
				return
			}
		}
	}
}

// ack acknowledges a message, removing it from the pending list.
func (c *Consumer) ack(ctx context.Context, streamID string) {
	err := c.client.XAck(ctx, c.cfg.InboundStream, c.cfg.ConsumerGroup, streamID).Err()
	if err != nil {
		log.Printf("[CONSUMER] XACK failed for %s: %v", streamID, err)
	}
}

// HealthCheck verifies Redis is reachable.
func (c *Consumer) HealthCheck(ctx context.Context) error {
	return c.client.Ping(ctx).Err()
}

// Close shuts down the Redis connection.
func (c *Consumer) Close() error {
	return c.client.Close()
}

// parseMessage converts a Redis stream message into our Message type.
func parseMessage(xmsg goredis.XMessage) (*Message, error) {
	dataStr, ok := xmsg.Values["data"].(string)
	if !ok {
		return nil, fmt.Errorf("missing or invalid 'data' field")
	}

	var msg Message
	if err := json.Unmarshal([]byte(dataStr), &msg); err != nil {
		return nil, fmt.Errorf("JSON unmarshal failed: %w", err)
	}

	msg.StreamID = xmsg.ID

	// Validate required fields
	if msg.MessageID == "" {
		return nil, fmt.Errorf("missing message_id")
	}
	if msg.SessionID == "" {
		return nil, fmt.Errorf("missing session_id")
	}

	return &msg, nil
}

File: beautifulplanet/safepaw/services/agent/src/consumer.ts
// =============================================================
// NOPEnclaw Agent — Redis Streams Consumer
// =============================================================
// Reads messages from nopenclaw_agent_inbox using XREADGROUP.
//
// Pattern mirrors the Router's Go consumer:
// 1. XREADGROUP with consumer groups for load distribution
// 2. XACK after successful processing
// 3. Batch processing with configurable concurrency
//
// If the Agent crashes, unacked messages stay in the pending
// list (PEL) and can be reclaimed by another instance.
// =============================================================

import Redis from "ioredis";
import { Config } from "./config";

/**
 * InboundMessage — what the Agent receives from the Router.
 * Matches the full inbound message format from the Gateway.
 * The Router forwards the complete message for maximum context.
 */
export interface InboundMessage {
  /** Redis stream entry ID — needed for XACK */
  streamId: string;
  /** Application-level message identifier */
  messageId: string;
  /** Session that sent the message — response routes back here */
  sessionId: string;
  /** Channel the message originated from (e.g., "discord", "telegram") */
  channel: string;
  /** Who sent the message */
  senderId: string;
  /** Platform of the sender */
  senderPlatform: string;
  /** MIME-like content type ("text/plain", "text/markdown", etc.) */
  contentType: string;
  /** The actual message content */
  content: string;
  /** Optional key-value metadata */
  metadata: Record<string, string>;
  /** Unix epoch seconds */
  timestamp: number;
}

/**
 * Handler processes a single inbound message.
 * Must throw on failure (message stays in PEL for retry).
 */
export type Handler = (msg: InboundMessage) => Promise<void>;

/**
 * Consumer reads from a Redis Stream using consumer groups.
 * Manages the read loop plus batch-level concurrency.
 */
export class Consumer {
  private client: Redis;
  private cfg: Config;
  private running = false;
  private reclaimerTimer: ReturnType<typeof setInterval> | null = null;

  constructor(cfg: Config) {
    this.cfg = cfg;

    // Parse host:port from addr string
    const [host, portStr] = cfg.redisAddr.split(":");
    const port = parseInt(portStr || "6379", 10);

    this.client = new Redis({
      host,
      port,
      password: cfg.redisPassword,
      db: cfg.redisDB,
      maxRetriesPerRequest: 3,
      lazyConnect: true,
      // Read timeout must exceed block time to prevent false disconnects.
      // XREADGROUP BLOCK blocks the connection for up to blockTimeMs.
      // If the socket timeout is shorter, ioredis kills the connection.
      commandTimeout: cfg.blockTimeMs + 5000,
    });
  }

  /**
   * Connect to Redis and ensure the consumer group exists.
   * Idempotent — safe to call even if the group already exists.
   */
  async connect(): Promise<void> {
    await this.client.connect();
    console.log(
      `[CONSUMER] Connected to ${this.cfg.redisAddr}, ` +
        `group=${this.cfg.consumerGroup}, consumer=${this.cfg.consumerName}`
    );

    // Create consumer group if it doesn't exist.
    // "0" means: start reading from the beginning of the stream.
    // MKSTREAM creates the stream if it doesn't exist yet.
    try {
      await this.client.xgroup(
        "CREATE",
        this.cfg.agentInboxStream,
        this.cfg.consumerGroup,
        "0",
        "MKSTREAM"
      );
      console.log(
        `[CONSUMER] Created group ${this.cfg.consumerGroup} on ${this.cfg.agentInboxStream}`
      );
    } catch (err: unknown) {
      // "BUSYGROUP" means group already exists — that's fine
      const message = err instanceof Error ? err.message : String(err);
      if (!message.includes("BUSYGROUP")) throw err;
      console.log(
        `[CONSUMER] Group ${this.cfg.consumerGroup} already exists (OK)`
      );
    }
  }

  /**
   * Run starts the read loop. Blocks until the AbortSignal fires.
   *
   * Processing model:
   * 1. XREADGROUP fetches up to batchSize messages
   * 2. All messages in the batch are processed concurrently
   * 3. Successful messages are XACKed individually
   * 4. Failed messages stay in the pending list for retry
   * 5. Loop repeats
   *
   * Concurrency is bounded by batchSize — at most batchSize
   * handler calls are in-flight at any time.
   */
  async run(handler: Handler, signal: AbortSignal): Promise<void> {
    this.running = true;
    console.log(
      `[CONSUMER] Read loop starting — batch=${this.cfg.batchSize}, ` +
        `block=${this.cfg.blockTimeMs}ms`
    );

    // Start pending message reclaimer (mirrors Router's pendingReclaimer).
    // Recovers messages stuck in PEL due to crashes or timeouts.
    this.startReclaimer(handler);

    while (this.running && !signal.aborted) {
      try {
        // XREADGROUP GROUP <group> <consumer> COUNT <batch> BLOCK <ms> STREAMS <stream> >
        const results = await this.client.xreadgroup(
          "GROUP",
          this.cfg.consumerGroup,
          this.cfg.consumerName,
          "COUNT",
          this.cfg.batchSize,
          "BLOCK",
          this.cfg.blockTimeMs,
          "STREAMS",
          this.cfg.agentInboxStream,
          ">"
        );

        // null means BLOCK timed out with no new messages
        if (!results) {
          console.log("[CONSUMER] XREADGROUP BLOCK timed out — no new messages (normal)");
          continue;
        }

        // ioredis xreadgroup returns: [[streamName, [[id, [field, value, ...]], ...]]]
        // TypeScript doesn't know the exact shape, so we cast it.
        const streams = results as [string, [string, string[]][]][];
        const batch: InboundMessage[] = [];

        for (const [, entries] of streams) {
          for (const [streamId, fields] of entries) {
            const msg = this.parseMessage(streamId, fields);
            if (!msg) {
              // ACK malformed messages so they don't clog the pending list forever.
              // Without this, unparseable entries stay in PEL permanently.
              await this.client.xack(
                this.cfg.agentInboxStream,
                this.cfg.consumerGroup,
                streamId
              );
              continue;
            }

            // Reject oversized messages at the consumer boundary
            if (msg.content.length > this.cfg.maxMessageSize) {
              console.error(
                `[CONSUMER] Message ${msg.messageId} too large ` +
                  `(${msg.content.length} bytes, max ${this.cfg.maxMessageSize}). ACKing to discard.`
              );
              await this.client.xack(
                this.cfg.agentInboxStream,
                this.cfg.consumerGroup,
                streamId
              );
              continue;
            }

            batch.push(msg);
          }
        }

        if (batch.length === 0) {
          console.log("[CONSUMER] Batch empty after filtering — skipping");
          continue;
        }

        console.log(`[CONSUMER] Processing batch of ${batch.length} messages`);

        // Process all messages in the batch concurrently.
        // Each handler call is independent — failure of one doesn't
        // affect others. Successful messages are XACKed individually.
        await Promise.allSettled(
          batch.map(async (msg) => {
            try {
              await handler(msg);
              // ACK on success — removes from pending list
              await this.client.xack(
                this.cfg.agentInboxStream,
                this.cfg.consumerGroup,
                msg.streamId
              );
              console.log(
                `[CONSUMER] ACKed msg=${msg.messageId} stream=${msg.streamId}`
              );
            } catch (err) {
              // Message stays in PEL — will be reclaimed on next
              // pending scan or by another consumer instance.
              console.error(
                `[CONSUMER] Handler failed for msg=${msg.messageId} ` +
                  `session=${msg.sessionId}: ${err}`
              );
            }
          })
        );
      } catch (err) {
        if (signal.aborted) break;
        console.error(`[CONSUMER] Read loop error: ${err}`);
        // Back off before retrying to avoid tight error loops
        await new Promise((r) => setTimeout(r, 1000));
      }
    }

    console.log("[CONSUMER] Read loop stopped");
  }

  /**
   * Parse a raw Redis stream entry into an InboundMessage.
   * The entry format is: ["data", "<json_string>"]
   * Returns null if the entry is malformed.
   */
  private parseMessage(
    streamId: string,
    fields: string[]
  ): InboundMessage | null {
    // Fields come as flat array: [key1, val1, key2, val2, ...]
    const dataIdx = fields.indexOf("data");
    if (dataIdx === -1 || dataIdx + 1 >= fields.length) {
      console.error(
        `[CONSUMER] Entry ${streamId} missing 'data' field — skipping`
      );
      return null;
    }

    try {
      const raw = JSON.parse(fields[dataIdx + 1]);
      return {
        streamId,
        messageId: raw.message_id || "",
        sessionId: raw.session_id || "",
        channel: raw.channel || "",
        senderId: raw.sender_id || "",
        senderPlatform: raw.sender_platform || "",
        contentType: raw.content_type || "",
        content: raw.content || "",
        metadata: raw.metadata || {},
        timestamp: raw.timestamp || 0,
      };
    } catch (err) {
      console.error(
        `[CONSUMER] Failed to parse entry ${streamId}: ${err}`
      );
      return null;
    }
  }

  // ================================================================
  // Pending Message Reclaimer — recovers stuck PEL messages
  // ================================================================
  // Mirrors the Router's Go pendingReclaimer. Without this, any
  // message that fails mid-processing (crash, timeout, OOM) stays
  // stuck in the pending entries list (PEL) forever.
  //
  // How it works:
  //   1. XPENDING scans for messages idle longer than ackTimeoutMs
  //   2. Messages under maxRetries → XCLAIM and re-process
  //   3. Messages over maxRetries → dead-letter to DLQ stream
  //   4. DLQ entries can be manually investigated / replayed
  // ================================================================

  /**
   * Start the periodic pending message reclaimer.
   * Runs on a setInterval — checks every ackTimeoutMs.
   */
  private startReclaimer(handler: Handler): void {
    const intervalMs = 30_000; // Check every 30s
    console.log(
      `[RECLAIMER] Started (interval=${intervalMs}ms, maxRetries=${this.cfg.maxRetries})`
    );

    this.reclaimerTimer = setInterval(async () => {
      if (!this.running) return;
      try {
        await this.reclaimPending(handler);
      } catch (err) {
        console.error(`[RECLAIMER] Error during reclaim cycle: ${err}`);
      }
    }, intervalMs);
  }

  /**
   * Stop the reclaimer timer.
   */
  private stopReclaimer(): void {
    if (this.reclaimerTimer) {
      clearInterval(this.reclaimerTimer);
      this.reclaimerTimer = null;
      console.log("[RECLAIMER] Stopped");
    }
  }

  /**
   * Scan the PEL for stuck messages and either re-process or dead-letter them.
   *
   * Uses XPENDING to find idle messages, then:
   * - If retryCount < maxRetries → XCLAIM + re-process
   * - If retryCount >= maxRetries → write to DLQ + XACK
   */
  private async reclaimPending(handler: Handler): Promise<void> {
    // XPENDING <stream> <group> IDLE <ms> - + <count>
    // Returns messages delivered but not yet acknowledged
    const idleMs = 30_000; // Messages idle for 30s+ are considered stuck

    // ioredis xpending with extended args
    const pending = await this.client.call(
      "XPENDING",
      this.cfg.agentInboxStream,
      this.cfg.consumerGroup,
      "IDLE",
      idleMs,
      "-",
      "+",
      this.cfg.batchSize
    ) as string[][];

    if (!pending || !Array.isArray(pending) || pending.length === 0) return;

    console.log(`[RECLAIMER] Found ${pending.length} stuck messages in PEL`);

    for (const entry of pending) {
      if (!this.running) return;

      // XPENDING extended returns: [messageId, consumer, idleTime, deliveryCount]
      const [msgId, , , deliveryCountStr] = entry;
      const deliveryCount = parseInt(String(deliveryCountStr), 10) || 0;

      // Dead-letter if retries exceeded
      if (deliveryCount >= this.cfg.maxRetries) {
        console.log(
          `[RECLAIMER] Message ${msgId} exceeded max retries ` +
            `(${deliveryCount}/${this.cfg.maxRetries}) — dead-lettering`
        );
        await this.deadLetter(msgId, deliveryCount);
        await this.client.xack(
          this.cfg.agentInboxStream,
          this.cfg.consumerGroup,
          msgId
        );
        continue;
      }

      // XCLAIM: take ownership and re-deliver
      try {
        const claimed = await this.client.xclaim(
          this.cfg.agentInboxStream,
          this.cfg.consumerGroup,
          this.cfg.consumerName,
          idleMs,
          msgId
        ) as unknown as [string, string[]][];

        if (!claimed || claimed.length === 0) continue;

        for (const [claimedId, fields] of claimed) {
          const msg = this.parseMessage(claimedId, fields);
          if (!msg) {
            console.log(
              `[RECLAIMER] Malformed reclaimed message ${claimedId} — ACKing to discard`
            );
            await this.client.xack(
              this.cfg.agentInboxStream,
              this.cfg.consumerGroup,
              claimedId
            );
            continue;
          }

          console.log(
            `[RECLAIMER] Re-processing msg=${msg.messageId} session=${msg.sessionId} ` +
              `(attempt #${deliveryCount + 1})`
          );

          try {
            await handler(msg);
            await this.client.xack(
              this.cfg.agentInboxStream,
              this.cfg.consumerGroup,
              msg.streamId
            );
            console.log(
              `[RECLAIMER] Successfully re-processed msg=${msg.messageId}`
            );
          } catch (err) {
            console.error(
              `[RECLAIMER] Re-processing failed for msg=${msg.messageId}: ${err}`
            );
            // Leave in PEL — will be picked up next cycle with incremented retry count
          }
        }
      } catch (err) {
        console.error(`[RECLAIMER] XCLAIM failed for ${msgId}: ${err}`);
      }
    }
  }

  /**
   * Write a failed message to the dead-letter queue (DLQ).
   * DLQ stream name = agent inbox stream + ":dlq" suffix.
   * Preserves original data for manual investigation / replay.
   */
  private async deadLetter(
    streamId: string,
    retryCount: number
  ): Promise<void> {
    const dlqStream = this.cfg.agentInboxStream + ":dlq";

    try {
      // Read original message data
      const msgs = await this.client.xrange(
        this.cfg.agentInboxStream,
        streamId,
        streamId
      );

      if (!msgs || msgs.length === 0) {
        console.error(
          `[DLQ] Could not read original message ${streamId} for dead-lettering`
        );
        return;
      }

      const [, fields] = msgs[0];
      const dataIdx = fields.indexOf("data");
      const originalData =
        dataIdx !== -1 && dataIdx + 1 < fields.length
          ? fields[dataIdx + 1]
          : "";

      // Write to DLQ with metadata
      await this.client.xadd(
        dlqStream,
        "MAXLEN",
        "~",
        "1000", // Cap DLQ — 1000 failed messages
        "*",
        "data",
        originalData,
        "original_id",
        streamId,
        "retry_count",
        String(retryCount),
        "consumer",
        this.cfg.consumerName,
        "dead_lettered_at",
        String(Math.floor(Date.now() / 1000))
      );

      console.log(
        `[DLQ] Message ${streamId} written to ${dlqStream} (retries=${retryCount})`
      );
    } catch (err) {
      console.error(
        `[DLQ] Failed to write message ${streamId} to DLQ: ${err}`
      );
    }
  }

  /**
   * Health check — verifies Redis connectivity.
   */
  async healthCheck(): Promise<void> {
    const result = await this.client.ping();
    if (result !== "PONG") {
      throw new Error(`Redis health check failed: ${result}`);
    }
  }

  /**
   * Gracefully stop the consumer.
   */
  stop(): void {
    console.log("[CONSUMER] Stop requested — finishing in-flight work");
    this.running = false;
    this.stopReclaimer();
  }

  /**
   * Close the Redis connection.
   */
  async close(): Promise<void> {
    this.running = false;
    this.stopReclaimer();
    await this.client.quit();
    console.log("[CONSUMER] Redis connection closed");
  }
}
